peer/peer_test.go

```diff
@@ -0,0 +1,384 @@
+// Copyright (c) 2015 The btcsuite developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package peer_test
+
+import (
+	"io"
+	"net"
+	"testing"
+	"time"
+
+	"github.com/btcsuite/btcd/addrmgr"
+	"github.com/btcsuite/btcd/blockchain"
+	"github.com/btcsuite/btcd/peer"
+	"github.com/btcsuite/btcd/wire"
+)
+
+// Mocks a network connection. Implements the net.Conn interface
+type Conn struct {
+	// Local network & address for the connection
+	LNet, LAddr string
+	// Remote network & address for the connection
+	RNet, RAddr string
+	// In messages will be written to this buffer
+	In io.Writer
+	// Out messages will be read from this buffer
+	Out io.Reader
+	// If non-nil, send value to it on close
+	closec chan bool
+}
+
+func (c *Conn) Read(b []byte) (n int, err error) {
+	return c.Out.Read(b)
+}
+
+func (c *Conn) Write(b []byte) (n int, err error) {
+	return c.In.Write(b)
+}
+
+func (c Conn) LocalAddr() net.Addr {
+	return &Addr{c.LNet, c.LAddr}
+}
+
+func (c Conn) RemoteAddr() net.Addr {
+	return &Addr{c.RNet, c.RAddr}
+}
+
+func (c *Conn) Close() error {
+	select {
+	case c.closec <- true:
+	default:
+	}
+	return nil
+}
+
+// Not implemented
+func (c Conn) SetDeadline(t time.Time) error      { return nil }
+func (c Conn) SetReadDeadline(t time.Time) error  { return nil }
+func (c Conn) SetWriteDeadline(t time.Time) error { return nil }
+
+// Mocks a network address
+type Addr struct {
+	Net, Addr string
+}
+
+func (m Addr) Network() string { return m.Net }
+func (m Addr) String() string  { return m.Addr }
+
+// pipe turns two mock connections into a full-duplex connection similar to
+// net.Pipe to allow pipe's with (fake) addresses.
+func pipe(c1, c2 *Conn) (*Conn, *Conn) {
+	r1, w1 := io.Pipe()
+	r2, w2 := io.Pipe()
+	c1.In = w1
+	c2.Out = r1
+	c1.Out = r2
+	c2.In = w2
+
+	return c1, c2
+}
+
+// addrMgr is the test address manager.
+// It is global so as to be accessible from mock listeners.
+var addrMgr = addrmgr.New("test", lookupFunc)
+
+var gotHandlers = make(chan string, 11)
+
+var wantHandlers = map[string]struct{}{
+	"handleVersionMsg":    struct{}{},
+	"handleMemPoolMsg":    struct{}{},
+	"handleTxMsg":         struct{}{},
+	"handleBlockMsg":      struct{}{},
+	"handleInvMsg":        struct{}{},
+	"handleHeadersMsg":    struct{}{},
+	"handleGetDataMsg":    struct{}{},
+	"handleGetBlocksMsg":  struct{}{},
+	"handleGetHeadersMsg": struct{}{},
+	"handleGetAddrMsg":    struct{}{},
+	"handleAddrMsg":       struct{}{},
+}
+
+func handleVersionMsg(p *peer.Peer, msg *wire.MsgVersion) {
+	addrMgr.AddAddress(p.NA(), p.NA())
```

Comment:
The intent of adding an address to the address manager is to add the remote address and the source address in order to help establish which ones are capable of communicating.  Since this is a test it's probably not a big deal to use the same one, but it definitely looks strange to see it here like this.